### ISSUE TITLE
runit can't work inside of Linux containers

### ISSUE THREAD

**Issue by unsystemizer** [CounterpartyXCP/federatednode#261]
- When run.py runs container stops and throws you out the moment `runit` is installed
- Upon further inspection it appears `runit` isn't compatible with containers

It seems a workaround would be required for run.py. Maybe add a "Are you installing inside of a Linux container (y/N)?" question which would on YES skip `runit` package installation.

Another option would be to ask folks to use https://github.com/phusion/baseimage-docker.
##

**Comment by unsystemizer** [comment]
Hmm, after intensive manipulation I get stuck at this apsw error. I do have python3-apsw installed.

```
  File "/home/xcp/federatednode_build/env/build/counterpartycli/setup.py", line 21, in run

    generate_config_files()

  File "/home/xcp/federatednode_build/env/build/counterpartycli/counterpartycli/setup.py", line 142, in generate_config_files

    from counterpartycli.server import CONFIG_ARGS as SERVER_CONFIG_ARGS

  File "/home/xcp/federatednode_build/env/build/counterpartycli/counterpartycli/server.py", line 8, in <module>

    from counterpartylib.lib import log

  File "/home/xcp/federatednode_build/env/lib/python3.4/site-packages/counterpartylib/lib/log.py", line 16, in <module>

    from counterpartylib.lib import util

  File "/home/xcp/federatednode_build/env/lib/python3.4/site-packages/counterpartylib/lib/util.py", line 7, in <module>

    import apsw

ImportError: No module named 'apsw'

----------------------------------------
Cleaning up...
  Removing temporary dir /home/xcp/federatednode_build/env/build...
Command /home/xcp/federatednode_build/env/bin/python3.4 -c "import setuptools, tokenize;__file__='/home/xcp/federatednode_build/env/build/counterpartycli/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-efkclla6-record/install-record.txt --single-version-externally-managed --compile --install-headers /home/xcp/federatednode_build/env/include/site/python3.4 failed with error code 1 in /home/xcp/federatednode_build/env/build/counterpartycli
Exception information:
Traceback (most recent call last):
  File "/home/xcp/federatednode_build/env/lib/python3.4/site-packages/pip/basecommand.py", line 122, in main
    status = self.run(options, args)
  File "/home/xcp/federatednode_build/env/lib/python3.4/site-packages/pip/commands/install.py", line 283, in run
    requirement_set.install(install_options, global_options, root=options.root_path)
  File "/home/xcp/federatednode_build/env/lib/python3.4/site-packages/pip/req.py", line 1435, in install
    requirement.install(install_options, global_options, *args, **kwargs)
  File "/home/xcp/federatednode_build/env/lib/python3.4/site-packages/pip/req.py", line 706, in install
    cwd=self.source_dir, filter_stdout=self._filter_install, show_stdout=False)
  File "/home/xcp/federatednode_build/env/lib/python3.4/site-packages/pip/util.py", line 697, in call_subprocess
    % (command_desc, proc.returncode, cwd))
pip.exceptions.InstallationError: Command /home/xcp/federatednode_build/env/bin/python3.4 -c "import setuptools, tokenize;__file__='/home/xcp/federatednode_build/env/build/counterpartycli/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-efkclla6-record/install-record.txt --single-version-externally-managed --compile --install-headers /home/xcp/federatednode_build/env/include/site/python3.4 failed with error code 1 in /home/xcp/federatednode_build/env/build/counterpartycli
```

**Comment by robby-dermody** [comment]
What do you mean by "container"? Docker?
Not using runit is a pretty big change. There are alternatives (we used upstart or init.d earlier I believe) but there'd have to be a pretty good reason to move to them.

**Comment by unsystemizer** [comment]
Yeah, I used Docker, but I assume the "standard" lxc also disallows it.

**Comment by robby-dermody** [comment]
Did you try https://github.com/phusion/baseimage-docker  ?

**Comment by unsystemizer** [comment]
Yeah I did that too, getting apsw not installed error, although it's installed. 
After 4 hours of trying (first I tried CW, then counterparty-server) I thought to start from something simple, such as `pip3 install counterparty-server` and the same for counterparty-client. Even that requires some poking around, but it installed kind of okay on `baseimage-docker`. I had to install some Python packages manually later (the installer missed them). 
I'll share my notes later (I am getting ready to run it now).

**Comment by unsystemizer** [comment]
## counterparty-server in LXC

- Start with the official Ubuntu 14.04 image
- Once inside:
apt-get update
apt-get install python3-apsw  python3-dev python3-pip wget zip gem ruby ruby-dev -y; gem install json
- Then inside of container change the timezone to UTC (recommended):
dpkg-reconfigure tzdata # set to UTC
pip3 install pytest prettytable # CP install skips these!
vim /etc/default/locale # may want to add it here
sudo locale -a # check
sudo locale-gen "en_US.UTF-8"
- Add these lines to /root/.bashrc (or for another user if you plan to use "xcp" or whatever UID):
export LANG=en_US.UTF-8
export LC_ALL="en_US.UTF-8"
- Then 
sudo dpkg-reconfigure locales
source /root/.bashrc
- Install CP
wget https://pypi.python.org/packages/source/r/requests/requests-2.9.1.tar.gz#md5=0b7f480d19012ec52bab78292efd976d
- Decompress and install requests 2.9.1
python3 setup.py install
- Check
pip3 list
- Install counterparty-cli:
wget https://github.com/CounterpartyXCP/counterparty-lib/archive/master.zip
unzip master.zip; rm -rf master.zip;
- Then build and install
- Repeat for the CLI
wget https://github.com/CounterpartyXCP/counterparty-cli/archive/master.zip
unzip master.zip; rm -rf master.zip;
- Now `counterparty-server --help` should work and after some cleanup we can commit the image as per usual Docker procedure

## Notes

- Installed this way, it works fine, but by default only one thing should run, which means if you run the server, you can't run the client, so I tested only the server (reparse, start, stop, etc)
- Needs more work, though: the image would have to be trimmed down, maybe build-essential and python3-dev stuff can be removed, apt cache cleaned, etc.
- EXPOSE and RUN should be added to YAML
- Container should be executed with "-v" to map counterparty.db from the outside (the host) - the container itself should be stateless (config file, data, etc. could/should all be outside).
- etc.

**Comment by robby-dermody** [comment]
interesting. thanks for trying that out. yes, as it is, the federated node setup script runs all services on a single box, which if taken out to containers should be changed to run a single service per container (as per best practice). With the phusion's `base-image`, one can run multiple services through, via `runit`, actually....they try to make a case that this is fine in many cases. Perhaps with a few small services, but probably not a good thing with all of the services that the federated node runs.

Have you seen the `-docker` repos at https://github.com/Dogeparty ?? They are a bit out of date but may serve as a good starting point to having a fully dockerized setup.

**Comment by robby-dermody** [comment]
Specifically, https://github.com/Dogeparty/dogecoind-docker/blob/master/Dockerfile may be of interest

**Comment by unsystemizer** [comment]
This is my first Docker image ever, I just looked it's 269 MB (on top of the official U14.04 base), certainly bloated but I just wanted to finish it this weekend so it is what it is. 
There's a concept of linked Docker images, which allows one to create a bunch of related and linked images with port mapping between them and so on, and run them as a group. I wonder if that may be easier to control (and maintain). 

Phushion's `base-image` is a great idea, but during the short time I tried to use it it seems the fiddling with the image has taken its toll - it's got some unique quirks compared to the official image, that's why I went with the official image in the end.

Even in the official image some pip3 requirements don't install, I wonder why is that.

I'll reference the Dockerfile next time. It'd be great to have everything in containers, eventually.
